src/PerfCounters.cc

```diff
@@ -268,29 +271,36 @@ static void do_branches() {
 #endif
 
 static void check_working_counters() {
-  struct perf_event_attr attr = rr::ticks_attr;
-  attr.sample_period = 0;
-  struct perf_event_attr attr2 = rr::cycles_attr;
-  attr.sample_period = 0;
-  ScopedFd fd = start_counter(0, -1, &attr);
-  ScopedFd fd2 = start_counter(0, -1, &attr2);
-  do_branches();
-  int64_t events = read_counter(fd);
-  int64_t events2 = read_counter(fd2);
+  int64_t events = 0;
+  int64_t events2 = 0;
+  for (const auto &perf_attr: perf_attrs) {
+    struct perf_event_attr attr = perf_attr.ticks;
+    attr.sample_period = 0;
+    struct perf_event_attr attr2 = perf_attr.cycles;
+    attr2.sample_period = 0;
+    ScopedFd fd = start_counter(0, -1, &attr);
+    ScopedFd fd2 = start_counter(0, -1, &attr2);
+    do_branches();
+    events += read_counter(fd);
+    events2 += read_counter(fd2);
+  }
```

Comment:
Why aren't we checking that events/events2 are nonzero for every counter?

Reply:
I guess I can initialize the fds as a vector here as well. I was a little too lazy to do that but I guess in principle if we are in the mode where we are not pinned, we could be migrated to exactly the wrong core when running the loops and got a false positive of 0 count on all counters...

Reply:
OK I think I'm missing the big picture of how this all works.

Are you saying that for ticks, there are two different hardware counters, one that counts properly on P cores and returns zero on E cores, and one that counts properly on E cores and returns zero on P cores?

Reply:
Yes that is correct. More precisely there are two set of counters. One set only counts event for the P core and one set only for the E core. I don't know exactly what hardware features they corresponds to but to userspace, the kernel abstract them as completely different event sources. See the "dynamic PMU" in the manpage of perf_event_open (and ignore the /cpu part of it since it doesn't exist on arm apparently...).

Reply:
OK, thanks, that needs to be documented where these vectors are introduced. Something like
```
  // For heterogenous cores, there can be multiple event sources --- one corresponding to each PMU type.
  // So we have one set of event attributes and one set of fds for each PMU type and count events across PMU types.
```
But ... how are interrupts going to work? In principle we'd want to interrupt when the sum of ticks across the two event sources reaches our interrupt threshold right? But I assume we can't do that?

Reply:
> But ... how are interrupts going to work? In principle we'd want to interrupt when the sum of ticks across the two event sources reaches our interrupt threshold right? But I assume we can't do that?

That's a good point. I've never tried unbound recording (though since rseq is used by glibc now that won't work at the moment anyway...).

For a single core this would not be a problem of course, though for a single core a single event type could probably also work as well. I did it this way to make sure unbound recording isn't completely ruled out and to not couple the pinning logic and the core type detection logic too much....

Reply:
The interrupt problem makes me thing that this whole "separate counters for each PMU type" setup is basically broken by design, but I suppose we're stuck with it :-(.

We don't really need interrupts for recording, but we do use them there and we definitely need them for replay. I think unbound replay with migrations across P/E cores must be broken.

Reply:
> The interrupt problem makes me thing that this whole "separate counters for each PMU type" setup is basically broken by design, but I suppose we're stuck with it :-(.

Actually I think this is something that could be fixed in the kernel so it might be worth bringing it up so the relevant kernel people are aware how broken their design is.

Reply:
> The interrupt problem makes me thing that this whole "separate counters for each PMU type" setup is basically broken by design, but I suppose we're stuck with it :-(.

I assume you, or someone else, may have better luck asking about those to the kernel. I have a hard time asking the reasons behind all the decisions made without being called wanting to put crap into the kernel.

Along that line, it'll also be good if the kernel support trapping access to the `cntvct_el0` system register (i.e. the arm version of rdtsc) and the ones covered by HWCAP_CPUID...

Reply:
> I assume you, or someone else, may have better luck asking about those to the kernel.

Maybe. The problem is I don't know all the details.

Reply:
I don't either.... and when I had no luck trying to ask... which I assume because I asked in the wrong way or at the wrong place even though I was directed to ask there = = ....

I do assume you know the requirement better though...

Reply:
Anyway for now I think we simply can't support situations where tracees can migrate from P to E cores or vice versa. So we are dependent on some kind of binding whether we want to be or not. Which means we only need one set of perf counters/fds actually.

Reply:
Well, the static data in PerfCounters probably needs to cover both event sources, so we need the attribute vector. But I think each `PerfCounters` instance needs to take an optional CPU-binding parameter indicating which CPU the instance is bound to. On M1, this parameter is non-optional --- i.e. binding is required. Then I guess we can look up the CPU's PMU type and create one set of fds, just for that type.

Reply:
OK. I think I can make the CPU binding part and the perf counter picking part talk to each other better.

You are right that the global static data part doesn't need to be aware of what we are binding to, and that's the part that would have broken the abstraction really badly. We only need to be aware of the right type we we initialize the counter, at which time we should know which, if any, core we are bound to.